bpo-29812: Improve testing of token and tokenize

ammaraskar opened the pull request:
Adds a token.py generation test, and a cross-check between the token and tokenize modules.

The generation test is akin to the current automated file generator tests in test_keyword and test_symbol. The newly added check in tokenize ensures that all the literal tokens in token such as ':', '...', '@=' have a corresponding entry within the EXACT_TOKEN_TYPES dict in tokenize.

ammaraskar committed:
bpo-29812: Improve testing of token and tokenize

Adds a token.py generation test, and a cross-check between the
token and tokenize modules.

The generation test is akin to the current automated file generator tests
in test_keyword and test_symbol. The newly added check in tokenize ensures
that all the literal tokens in token such as ':', '...', '@=' have a
corresponding entry within the EXACT_TOKEN_TYPES dict in tokenize.

ammaraskar commented:
Weirdly enough this fails on AppVeyor (which is Windows, I believe). I'm not sure if this is something wrong with the test or an actual problem on windows, I'll look into it tomorrow when I have access to a windows computer.

berkerpeksag commented on Lib/test/test_token.py:
```diff
@@ -0,0 +1,52 @@
+import token
+import unittest
+from test import support
```
Please group stdlib modules together and sort them alphabetically.

berkerpeksag commented on Lib/test/test_token.py:
```diff
@@ -0,0 +1,52 @@
+import sys
+import subprocess
+
+TOKEN_FILE               = support.findfile('token.py')
```
Style nit: Don't add additional spaces before `=`.

berkerpeksag commented on Lib/test/test_token.py:
```diff
@@ -0,0 +1,52 @@
+        rc, stderr = self._generate_tokens(os.devnull, NONEXISTENT_FILE)
+        self.assertNotEqual(rc, 0)
+        self.assertIn(b'I/O error', stderr)
+        self.assertIn(bytes(NONEXISTENT_FILE, encoding='ascii'), stderr)
```
`NONEXISTENT_FILE` doesn't contain any non-ASCII chars. Do we really need `encoding='ascii'`?

Also, I prefer to have a `nonexistent_file` local variable in the test instead of a constant.

ammaraskar replied:
The encoding is required when passing a string in python3: https://docs.python.org/3/library/functions.html#bytes
https://docs.python.org/3/library/functions.html#bytearray

The reason its a b'string' is because it needs to be passed as an argument to Popen, changing it to a local might result in an easier fix for this though.

berkerpeksag commented on Lib/test/test_token.py:
```diff
@@ -0,0 +1,52 @@
+        self.assertIn(bytes(NONEXISTENT_FILE, encoding='ascii'), stderr)
+
+
+if __name__ == "__main__":
```
Style use: Please use single quotes where possible.

ammaraskar replied:
Will do, for reference, this was copied from https://github.com/python/cpython/blob/master/Lib/test/test_keyword.py

berkerpeksag commented on Lib/test/test_token.py:
```diff
@@ -0,0 +1,52 @@
+        stderr = proc.communicate()[1]
+        return proc.returncode, stderr
+
+    @unittest.skipIf(not os.path.exists(TOKEN_FILE),
```
I'd replace

```py
@unittest.skipIf(not condition)
```

with

```py
@unittest.skipUnless(condition)
```

berkerpeksag commented on Lib/test/test_token.py:
```diff
@@ -0,0 +1,52 @@
+                     'test only works from source build directory')
+    def test_real_token_file(self):
+        self._copy_file_without_generated_tokens(TOKEN_FILE, TEST_PY_FILE)
+        self.addCleanup(support.unlink, TEST_PY_FILE)
```
You can probably move the `addCleanup` call to inside of `_copy_file_without_generated_tokens` method.

berkerpeksag commented on Lib/test/test_token.py:
```diff
@@ -0,0 +1,52 @@
+                                                         TEST_PY_FILE))
+        self.assertTrue(filecmp.cmp(TOKEN_FILE, TEST_PY_FILE))
+
+    def test_missing_output_file_causes_Error(self):
```
Error -> error or exception

berkerpeksag commented on Lib/test/test_tokenize.py:
```diff
@@ -1349,6 +1349,25 @@ def assertExactTypeEqual(self, opstr, *optypes):
                          token.tok_name[token.ENDMARKER])
 
+    def test_all_literal_tokens(self):
+        NON_LITERALS = {token.ENDMARKER, token.NAME, token.NUMBER,
```
NON_LITERALS -> non_literals or non_literal_tokens

berkerpeksag replied:
Also, I find hanging indents much more readable and they will make future diffs less noisy:

```py
non_literal_tokens = {
    token.STRING, token.NEWLINE, token.INDENT,
    ...,
    # You can add another line without having to modify the '}' so diffs will be less noisy.
    token.N_TOKENS, token.NT_OFFSET,
}
```

berkerpeksag commented on Lib/test/test_tokenize.py:
```diff
@@ -1349,6 +1349,25 @@ def assertExactTypeEqual(self, opstr, *optypes):
+                continue
+            name = tok_name[tok]
+            self.assertIn(tok, EXACT_TOKEN_TYPES.values(),
+                    "Literal token " + name + " not in EXACT_TOKEN_TYPES")
```
This is OK, but an alternative solution would be to use `self.subTest()`.

ammaraskar committed:
Changes for berker's review

ammaraskar commented:
Oh also, I tried to look into what was calling the AppVeyor failure when comparing the files. From what I can tell its because os.stat is returning an incorrect value for st_size on the test environment or something https://ci.appveyor.com/project/ammaraskar/cpython/build/1.0.9#L3949

Doing a file content comparison causes it to pass.

Nevermind, I don't think this is an incorrect return since the offset of the size is exactly the number of lines, I think this has to do with line endings.

ammaraskar committed:
Preserve newline conventions for skeleton file

Currently this code does not work on windows
when the skeleton file has unix line endings
because fp.write('\n') will have the LF
changed to CRLF due to python's newline
interpolation.

ammaraskar committed:
Add newly added non literal tokens to test

ammaraskar commented:
Hey @berkerpeksag, could you take a look at this, I updated it fixing the concerns in your review.

@bitdancer could I also get a review from you seeing as you made the issue on the tracker for this?

berkerpeksag commented on Lib/test/test_token.py:
```diff
@@ -0,0 +1,52 @@
+import os
+import subprocess
+import sys
+from test import support
```
Please split stdlib and non-stdlib modules.

berkerpeksag commented on Lib/test/test_token.py:
```diff
@@ -0,0 +1,52 @@
+        self.addCleanup(support.unlink, dest_file)
+
+    def _generate_tokens(self, skeleton_file, target_token_py_file):
+        proc = subprocess.Popen([sys.executable,
```
Can't we use the new `subprocess.run()` interface?

berkerpeksag commented on Lib/token.py:
```diff
@@ -129,33 +129,37 @@ def _main():
     with fp:
-        format = fp.read().split("\n")
+        format = fp.readlines()
+    nl = format[0][len(format[0].strip()):] if format else '\n'
```
I'd use more descriptive name like `newline`.

berkerpeksag commented on Lib/test/test_tokenize.py:
```diff
@@ -1351,6 +1351,28 @@ def assertExactTypeEqual(self, opstr, *optypes):
+                continue
+            name = tok_name[tok]
+            with self.subTest(name=name):
+                self.assertIn(tok, EXACT_TOKEN_TYPES.values())
```
Looking at [CI logs](https://travis-ci.org/python/cpython/jobs/246655056) of an earlier commit, I wonder if we should make the following message clearer:

```
======================================================================
FAIL: test_all_literal_tokens (test.test_tokenize.TestTokenize) (name='NL')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_tokenize.py", line 1368, in test_all_literal_tokens
    self.assertIn(tok, EXACT_TOKEN_TYPES.values())
AssertionError: 58 not found in dict_values([7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 52, 51, 49, 50])
```

berkerpeksag reviewed (commented):
Sorry for my late response. I left some comments. Please add a NEWS entry in `Misc/NEWS.d/next/Tests`.

Out of curiosity, how did you notice 3d4a043dbee837723748541b3298fef2d9c2b938? It looks like AppVeyor didn't [catch it](https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.3738).

ammaraskar commented:
Thanks for your review @berkerpeksag. Closing this as per discussion on bpo.
